src/test/java/io/vertx/core/http/HttpBandwidthLimitingTest.java

```diff
@@ -0,0 +1,289 @@
+import org.apache.commons.lang.time.StopWatch;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+import org.junit.runners.Parameterized.Parameters;
+
+import io.netty.handler.codec.http.HttpHeaderNames;
+import io.vertx.core.Handler;
+import io.vertx.core.Vertx;
+import io.vertx.core.buffer.Buffer;
+import io.vertx.test.core.TestUtils;
+
+@RunWith(Parameterized.class)
+public class HttpBandwidthLimitingTest extends Http2TestBase {
+  private static final int OUTBOUND_LIMIT = 2 * 64 * 1024;  // 64KB/s
+  private static final int INBOUND_LIMIT = 64 * 1024;   // 64KB/s
+  private static final int TEST_CONTENT_SIZE = 64 * 1024 * 4;   // 64 * 4 = 256KB
+
+  private final File sampleF = new File(new File(TestUtils.MAVEN_TARGET_DIR, "test-classes"), "test_traffic.txt");
+  private final Handlers HANDLERS = new Handlers();
+
+  @Parameters(name = "HTTP {0}")
+  public static Iterable<Object[]> data() {
+
+    Function<Vertx, HttpServer> http1ServerFactory = (v) -> Providers.http1Server(v, INBOUND_LIMIT, OUTBOUND_LIMIT);
+    Function<Vertx, HttpServer> http2ServerFactory = (v) -> Providers.http2Server(v, INBOUND_LIMIT, OUTBOUND_LIMIT);
+    Function<Vertx, HttpClient> http1ClientFactory = (v) -> v.createHttpClient();
+    Function<Vertx, HttpClient> http2ClientFactory = (v) -> v.createHttpClient(createHttp2ClientOptions());
+
+    return Arrays.asList(new Object[][] {
+      { 1.1, http1ServerFactory, http1ClientFactory },
+      { 2.0, http2ServerFactory, http2ClientFactory }
+    });
+  }
+
+  private Function<Vertx, HttpServer> serverFactory;
+  private Function<Vertx, HttpClient> clientFactory;
+
+  public HttpBandwidthLimitingTest(double protoVersion, Function<Vertx, HttpServer> serverFactory,
+                                   Function<Vertx, HttpClient> clientFactory) {
+    this.serverFactory = serverFactory;
+    this.clientFactory = clientFactory;
+  }
+
+  @Before
+  public void setUp() throws Exception {
+    super.setUp();
+    server = serverFactory.apply(vertx);
+    client = clientFactory.apply(vertx);
+  }
+
+  @After
+  public void after() throws InterruptedException
+  {
+    CountDownLatch waitForClose = new CountDownLatch(1);
+    vertx.close().onComplete(resp -> waitForClose.countDown());
+    awaitLatch(waitForClose);
+  }
+
+  @Test
+  public void sendBufferThrottled() throws Exception {
+    Buffer expectedBuffer = TestUtils.randomBuffer(TEST_CONTENT_SIZE);
+
+    HttpServer testServer = serverFactory.apply(vertx);
+    StopWatch watch = new StopWatch();
+    testServer.requestHandler(HANDLERS.bufferRead(expectedBuffer));
+    startServer(testServer);
+
+    watch.start();
+    HttpClient testClient = clientFactory.apply(vertx);
+    read(expectedBuffer, testServer, testClient);
+    await();
+    watch.stop();
+
+    assertTrue(watch.getTime() > expectedTimeMillis(TEST_CONTENT_SIZE, OUTBOUND_LIMIT));
```

Comment:
we need here to use `Assert.assertTrue(...)` instead of `assert(...)` because `await()` was called before and the async test supposes the test is finished and no error can be reported